### ISSUE TITLE
Heroic unknown protocol error.

### ISSUE THREAD

**Issue by onurbbr** [Heroic-Games-Launcher/HeroicGamesLauncher#577]
Hello there. I had added my shortcuts from Heroic to the start menu. I get this type of error when I try to open a game from the start menu. The application does not open with the start menu and the shortcut on the desktop. Is this a bug or a problem with xdg-open?

![Screenshot_20210729_144239](https://user-images.githubusercontent.com/47495526/127486269-634c727e-16de-48cb-a3ff-b47e3fe0c26a.png)

Distro: Arch Linux
DE: KDE Plasma 5.22.4 (x11)
Kernel: Linux Zen 5.13.5
Heroic Version: 1.9.0 Mihawk

**Comment by dragonDScript** [comment]
I am investigating this issue. Does this happen with other ElectronJS apps (e.g. try to join to a discord server with a `discord.gg` url and see if it opens the desktop app).

**Comment by onurbbr** [comment]
> I am investigating this issue. Does this happen with other ElectronJS apps? (e.g. try to join to a discord server with a `discord.gg` url and see if it opens the desktop app).
> 
> By the way, can you reproduce this after an update & a reboot?

No problem on Discord. With the link, the Discord application opens directly and I can enter the group. I restarted my computer after the update but I still get the error.

**Comment by onurbbr** [comment]
Error as in the video.

https://user-images.githubusercontent.com/47495526/127534195-f8fa088f-6b47-4477-a023-9c5b1c9eb3a5.mp4

**Comment by flavioislima** [comment]
This is not an error with heroic this is an error with the KDE Plasma that is not allowing Heroic to register the protocol.
You can check this running heroic from the terminal. You will see that heroic will try to register the protocol but it will give an error. Probably your user doesn't have permission to change that, for some reason.

**Comment by flavioislima** [comment]
Just did another test. I removed the protocol `heroic://` from the handlers and opened heroic again, it registered the protocol and the apps were launching normally from shortcuts. so I think you're missing something in your system.

**Comment by onurbbr** [comment]
> This is not an error with heroic this is an error with the KDE Plasma that is not allowing Heroic to register the protocol.
> You can check this running heroic from the terminal. You will see that heroic will try to register the protocol but it will give an error. Probably your user doesn't have permission to change that, for some reason.

What kind of permission is needed for this?
The groups my user is attached to are as follows:
onur@archlinux   ~ ❯ cat /etc/group | grep onur                                                                                                                                                                           at  08:44:00 PM
sys:x:3:bin,onur
ftp:x:11:onur
mail:x:12:onur
log:x:19:onur
smmsp:x:25:onur
proc:x:26:polkitd,onur
games:x:50:onur
scanner:x:96:onur
power:x:98:onur
adm:x:999:daemon,onur
wheel:x:998:onur
utmp:x:997:onur
audio:x:996:brltty,onur
disk:x:995:onur
input:x:994:brltty,onur
kmem:x:993:onur
kvm:x:992:onur
lp:x:991:cups,onur
optical:x:990:onur
storage:x:987:onur
tty:x:5:brltty,onur
uucp:x:986:brltty,onur
video:x:985:onur,sddm
systemd-journal:x:983:onur
rfkill:x:982:onur
http:x:33:onur
dbus:x:81:onur
onur:x:1000:
brlapi:x:968:brltty,onur
docker:x:967:onur
libvirt:x:966:onur
realtime:x:965:onur
vboxusers:x:108:onur
wireshark:x:150:onur

**Comment by onurbbr** [comment]
I solved my problem. I entered the command as in the photo following the protocols in the [link](https://wiki.archlinux.org/title/XDG_MIME_Applications#mimeapps.list)  below and deleted the heroic entries in "~/.config/mimeapps.list". When I try to open the game again, heroic opens and then the game opens.
![Screenshot_20210729_211826](https://user-images.githubusercontent.com/47495526/127544903-e993a727-c465-45a0-9ee1-3a4f8f48f19d.png)
Thank you for your help.

https://user-images.githubusercontent.com/47495526/127545968-cf200ad8-76d0-4b03-95d1-67a8e592ca7a.mp4

**Comment by flavioislima** [comment]
Perfect, good job! Good to know that it worked :)

**Comment by onurbbr** [comment]
Edit: ![Screenshot_20210729_211826](https://user-images.githubusercontent.com/47495526/127544903-e993a727-c465-45a0-9ee1-3a4f8f48f19d.png)
After doing the operation in the picture, adding "x-scheme-handler/heroic=heroic.desktop" under the "Default Applications" section, the problem disappears.

**Comment by GitThisBugOff** [comment]
> Edit: ![Screenshot_20210729_211826](https://user-images.githubusercontent.com/47495526/127544903-e993a727-c465-45a0-9ee1-3a4f8f48f19d.png)
> After doing the operation in the picture, adding "x-scheme-handler/heroic=heroic.desktop" under the "Default Applications" section, the problem disappears.

Despite doing this I still get this problem in Kubuntu.

**Comment by KuleRucket** [comment]
The first game I added the shortcut for adding the following to .config/mimeapps.list:
`x-scheme-handler/heroic=<Game>.desktop`
I changed this to the following:
`x-scheme-handler/heroic=heroic.desktop`
Now all links work fine

**Comment by flavioislima** [comment]
@KuleRucket you're saying that heroic registered a new protocol after the game install?
I cannot see how it would do it if the way we register protocol is hardcoded for the string 'heroic' and the desktops are just a command to write a file with the launch information both on desktop and on the applications folder.
I have no idea why it have created the other scheme 🤔

**Comment by KuleRucket** [comment]
I installed the game Control via heroic, then added a desktop icon. The line added to .config/mimeapps.list was:
x-scheme-handler/heroic=Control.desktop

Clicking on the desktop icon raised the error message in the first report.
I manually corrected this to:
x-scheme-handler/heroic=heroic.desktop
And it launches fine. Also any further desktop icons added after this also work fine.

Some details FYI:
Distro: Arch (up to date as of today)
Desktop: KDE Plasma 5.23.3
Heroic: 1.10.3 (installed via AUR heroic-games-launcher-bin).

**Comment by flavioislima** [comment]
Ok. I'll do more investigation on that and try to fix it for good.

**Comment by ImperatorStorm** [comment]
can still reproduce, despite above fixes.

**Comment by GitThisBugOff** [comment]
This is what I get when executing the xdg-open manually through a terminal.

![Screenshot_20220204_192156](https://user-images.githubusercontent.com/39744079/152616116-0394ef89-9a34-4091-82fc-a9e5efc92095.png)

**Comment by gnoumph** [comment]
Hello,

Seems like this issue still occurs, at least for me!

When I start Heroic, an entry is added to my `~/.config/mimeapps.list`, but strange thing, it's `x-scheme-handler/heroic=PC Building Simulator.desktop`.  
If I replace with `x-scheme-handler/heroic=heroic.desktop`, everything works fine.

I'm using Kubuntu 21.10 with Heroic Launcher 2.2.6 (DEB version).

**Comment by Lifeline684** [comment]
Anyone able to make a video explaining how to fix? I got confused trying to do this and found no file with the name mimeapps.list in the .config folder. I’m really confused and new to Linux so I’d appreciate the help figuring out what to do!

**Comment by gnoumph** [comment]
> Anyone able to make a video explaining how to fix? I got confused trying to do this and found no file with the name mimeapps.list in the .config folder. I’m really confused and new to Linux so I’d appreciate the help figuring out what to do!

Hello,

* Open a file explorer
* Go to your personal folder (`~/` in previous message means "home directory of current user")
* Press `Ctrl` + `H` to display hidden files (those starting with a dot)
* Go to `.config/` folder
* Open `mimeapps.list` file with e text editor
* Edit the file, replacing `x-scheme-handler/heroic=A Game.desktop` with `x-scheme-handler/heroic=heroic.desktop`

If you do not have this file tell us your Linux and Heroic version.

**Comment by flavioislima** [comment]
This should be fixed on 2.3.0 I believe.

**Comment by Lifeline684** [comment]
> > Anyone able to make a video explaining how to fix? I got confused trying to do this and found no file with the name mimeapps.list in the .config folder. I’m really confused and new to Linux so I’d appreciate the help figuring out what to do!
> 
> Hello,
> 
> * Open a file explorer
> * Go to your personal folder (`~/` in previous message means "home directory of current user")
> * Press `Ctrl` + `H` to display hidden files (those starting with a dot)
> * Go to `.config/` folder
> * Open `mimeapps.list` file with e text editor
> * Edit the file, replacing `x-scheme-handler/heroic=A Game.desktop` with `x-scheme-handler/heroic=heroic.desktop`
> 
> If you do not have this file tell us your Linux and Heroic version.

checked and saw no file. my heroic version is 2.2.6 Oden and Linux version Linux 5.23.0-valve10.3-1neptune-02176-g5fe416c4acd8 (i think)

**Comment by Trixxle** [comment]
I also haven't been able to fix this issue and I also don't have a mineapps.list file that directory. I do have a mimeapps.list file in /.local/share/applicatios but the mimeapps.list file is empty. My Heroic version is 2.3.5 running SteamOS 3.2 stable

**Comment by flavioislima** [comment]
@Trixxle this is a different issue. Right now desktop shortcuts don't work for the Flatpak version of Heroic, that's the version Steam OS uses. 
The shortcuts are created but the protocol is never reached.

**Comment by dragonDScript** [comment]
> @Trixxle this is a different issue. Right now desktop shortcuts don't work for the Flatpak version of Heroic, that's the version Steam OS uses. The shortcuts are created but the protocol is never reached.

Is flatpak isolated like e.g. snap? Maybe that's the issue.

**Comment by DeaTh-G** [comment]
> @Trixxle this is a different issue. Right now desktop shortcuts don't work for the Flatpak version of Heroic, that's the version Steam OS uses. The shortcuts are created but the protocol is never reached.

Has this issue been resolved/look into yet? Using Heroic v2.4.3 from Flatpak on a Steam Deck still produces the same `Unable to create io-slave. klauncher said: Unknown protocol 'heroic'` error when trying to launch shortcuts.
The following is present in `~/.config/mimeapps.list`: `x-scheme-handler/heroic=com.heroicgameslauncher.hgl.desktop` installed via the following command: `xdg-mime default com.heroicgameslauncher.hgl.desktop x-scheme-handler/heroic`.

**Comment by dragonDScript** [comment]
> > @Trixxle this is a different issue. Right now desktop shortcuts don't work for the Flatpak version of Heroic, that's the version Steam OS uses. The shortcuts are created but the protocol is never reached.
> 
> Has this issue been resolved/look into yet? Using Heroic v2.4.3 from Flatpak on a Steam Deck still produces the same `Unable to create io-slave. klauncher said: Unknown protocol 'heroic'` error when trying to launch shortcuts. The following is present in `~/.config/mimeapps.list`: `x-scheme-handler/heroic=com.heroicgameslauncher.hgl.desktop` installed via the following command: `xdg-mime default com.heroicgameslauncher.hgl.desktop x-scheme-handler/heroic`.

I believe there was a workaround or a fix. I don't know

**Comment by skunkworxdark** [comment]
> > > @Trixxle this is a different issue. Right now desktop shortcuts don't work for the Flatpak version of Heroic, that's the version Steam OS uses. The shortcuts are created but the protocol is never reached.
> > 
> > 
> > Has this issue been resolved/look into yet? Using Heroic v2.4.3 from Flatpak on a Steam Deck still produces the same `Unable to create io-slave. klauncher said: Unknown protocol 'heroic'` error when trying to launch shortcuts. The following is present in `~/.config/mimeapps.list`: `x-scheme-handler/heroic=com.heroicgameslauncher.hgl.desktop` installed via the following command: `xdg-mime default com.heroicgameslauncher.hgl.desktop x-scheme-handler/heroic`.
> 
> I believe there was a workaround or a fix. I don't know

I am also getting this same issue on my steam deck.  I have looked through the troubleshooting guide but hasn't helped much.

**Comment by hylianux** [comment]
These workarounds do not seem to work, or if they did before, they don't work anymore.

Tried making a symbolic link from ~./config/mimeapps.list to ~/.local/share/applications/mimeapps.list

Tried editing mimeapps.list
  - deleting the heroic entry from the list entirely and making it regenerate it.
  - changing the heroic entry from `x-scheme-handler/heroic=com.heroicgameslauncher.hgl.desktop` to `x-scheme-handler/heroic=heroic.desktop`

Tried rebooting (because I know someone's going to ask haha).

Tried running it directly from the command line and it pops up with a very similar error message, I imagine the differences are due to updates in the past 3 years that have renamed things, but here it is nonetheless

<img width="452" height="159" alt="Image" src="https://github.com/user-attachments/assets/88d0066e-03db-4302-8781-3fe1bf381f54" />

what's interesting to note here, however, is this: 
```shell
xdg-open 'heroic://launch?appName=1097893768&runner=gog'                                                                                                                                      
kf.kio.core: "/usr/share/applications/com.heroicgameslauncher.hgl.desktop" contains supported protocols but doesn't use %u or %U in its Exec line! This is inconsistent.
```
After some [google-fu](https://specifications.freedesktop.org/desktop-entry-spec/latest/exec-variables.html), I went to the file in question `/usr/share/applications/com.heroicgameslauncher.hgl.desktop` and discovered what the problem was: the Exec line.
```
Exec=heroic
```
You need to change that to 
```
Exec=heroic %u
```

So, in summary: in 2025 the fix has nothing to do with modifying the mimeapps.list or making symlinks...

You have to modify the desktop entry's Exec line to "heroic %u".

Tried on Heroic Version 2.17.2

**Comment by ngmoviedo** [comment]
> You have to modify the desktop entry's Exec line to "heroic %u".

This worked perfectly for me.
Also in version 2.17.2.
Under Manjaro KDE.

**Comment by hylianux** [comment]
Version update to 2.18.1 changes it back, and I have to modify the desktop entry again.  

Noting this here just to point out that every time there's a new version of the heroic game launcher, I have to fix the desktop file.  Should I open a new issue for this?  

Does anyone foresee any problems as a result of this change?

Edit: I'd fix the problem myself, but I'm not exactly sure where.  I took a look at the sourcecode, and everywhere that references the desktop entry correctly establishes the "%u" part of the `Exec=` line... 

I'm on manjaro, an arch-based distro, so i checked the latest release's .pacman file, and it also exists there too.

Somewhere this `%u` parameter is getting dropped.  

I took a look at manjaro's repositories, downloaded the latest copy of HGL from them (2.18.1-2) and extracted it out - aha, found the culprit.  

This side of things is over my head, because I've never submitted a package to a distro's official repo's before... I'm not sure if this is something the heroic team maintains, or if there's some partnership deal, or if this is all magically handled by github... does someone from the manjaro team usually manage this?  Anyway - that's where the file is wrong, and will continue to be wrong with each new update, forcing all of us to keep fixing it.  

I suppose the best workaround here for us Arch-based people is to use the AUR version instead of any official repo version.

TL;DR - would love to fix this myself, but the problem as far as I can see is limited to manjaro's copy of the program, and I would need some guidance on how to fix manjaro's copy.

**Comment by ngmoviedo** [comment]
> Version update to 2.18.1 changes it back, and I have to modify the desktop entry again.

That's strange. I'm also on Manjaro and the update didn't revert that change for me. I haven't observed that behavior again.

**Comment by Darukineo** [comment]
I made it work with "usr/bin/heroic" 
I use a biglinux, based on Manjaro. hope this helps

**Comment by xihale** [comment]
Hand write a protocol handler can also solve this problem.

Edit `~/.local/share/applications/heroic-protocol.desktop`

```
[Desktop Entry]
Name=Heroic Protocol Handler
Comment=Handle heroic:// protocol links
Exec=/usr/bin/heroic %u
Icon=heroic
Terminal=false
Type=Application
MimeType=x-scheme-handler/heroic;
NoDisplay=true
Categories=Game;
```

And then flush the mime database & register x-scheme-handler/heroic to heroic-protocol.desktop

```
update-desktop-database ~/.local/share/applications
xdg-mime default heroic-protocol.desktop x-scheme-handler/heroic
```

Check whether done

```
xdg-mime query default x-scheme-handler/heroic
```

**Comment by hylianux** [comment]
> Hand write a protocol handler can also solve this problem.
> 
> Edit `~/.local/share/applications/heroic-protocol.desktop`
> 
> ```
> [Desktop Entry]
> Name=Heroic Protocol Handler
> Comment=Handle heroic:// protocol links
> Exec=/usr/bin/heroic %u
> Icon=heroic
> Terminal=false
> Type=Application
> MimeType=x-scheme-handler/heroic;
> NoDisplay=true
> Categories=Game;
> ```
> 
> And then flush the mime database & register x-scheme-handler/heroic to heroic-protocol.desktop
> 
> ```
> update-desktop-database ~/.local/share/applications
> xdg-mime default heroic-protocol.desktop x-scheme-handler/heroic
> ```
> 
> Check whether done
> 
> ```
> xdg-mime query default x-scheme-handler/heroic
> ```

Simple enough, works like a charm.  Can confirm.
